### PR TITLE
add demo and codes optimization

### PR DESCRIPTION
**Pull request by songchengjiang**

### CODE REVIEW

File: cocos/navmesh/CCNavMesh.cpp [hunk 1]
```diff
@@ -82,18 +118,38 @@ NavMesh::~NavMesh()
...
+    if (!loadGeomFile()) return false;
+    if (!loadNavMeshFile()) return false;
+
+    return true;
+}
+
+bool NavMesh::loadNavMeshFile()
+{
+    std::string fullPath = FileUtils::getInstance()->fullPathForFilename(_navFilePath);
     FILE* fp = fopen(fullPath.c_str(), "rb");
```
Comment (super626) [inline cocos/navmesh/CCNavMesh.cpp]: FileUtils::getInstance()->getStringFromFile

File: cocos/navmesh/CCNavMesh.cpp [hunk 2]
```diff
@@ -176,6 +232,190 @@ bool NavMesh::read()
     return true;
 }
 
+bool NavMesh::loadGeomFile()
+{
+    std::string fullPath = FileUtils::getInstance()->fullPathForFilename(_geomFilePath);
+    char* buf = 0;
+    FILE* fp = fopen(fullPath.c_str(), "rb");
```
Comment (super626) [inline cocos/navmesh/CCNavMesh.cpp]: FileUtils::getInstance()->getStringFromFile(

File: cocos/navmesh/CCNavMeshAgent.h [hunk 3]
```diff
@@ -86,15 +87,14 @@ class CC_DLL NavMeshAgent : public Component
         NODE_AND_NODE = NODE_TO_AGENT | AGENT_TO_NODE,
     };
 
-    typedef std::function<void(const NavMeshAgent *, bool &)> MoveCallback;
+    typedef std::function<void(NavMeshAgent *agent, float totalTimeAfterMove)> MoveCallback;
 
     static NavMeshAgent* create(const NavMeshAgentParam &param);
+    static std::string& getNavMeshAgentComponentName();
```
Comment (super626) [inline cocos/navmesh/CCNavMeshAgent.h]: static const std::string& getNavMeshAgentComponentName();

File: cocos/navmesh/CCNavMeshDebugDraw.cpp [hunk 4]
```diff
@@ -117,11 +124,13 @@ GLenum NavMeshDebugDraw::getPrimitiveType(duDebugDrawPrimitives prim)
     }
 }
 
-void cocos2d::NavMeshDebugDraw::drawImplement(const cocos2d::Mat4& transform, uint32_t flags)
+void NavMeshDebugDraw::drawImplement(const cocos2d::Mat4& transform, uint32_t flags)
 {
     _program->use();
     _program->setUniformsForBuiltins(transform);
     glEnable(GL_DEPTH_TEST);
+    glEnable(GL_CULL_FACE);
```
Comment (super626) [inline cocos/navmesh/CCNavMeshDebugDraw.cpp]: backup the opengl state and restore it after finish drawing

File: cocos/navmesh/CCNavMesh.h [hunk 5]
```diff
@@ -54,7 +54,7 @@ class CC_DLL NavMesh : public Ref
 {
 public:
 
-    static NavMesh* create(const std::string &filePath);
+    static NavMesh* create(const std::string &navFilePath, const std::string &geomFilePath);
```
Comment (super626) [inline cocos/navmesh/CCNavMesh.h:57]: NavMesh needs ray cast

File: cocos/navmesh/CCNavMesh.cpp (hunk 1 above)
Comment (songchengjiang) [inline cocos/navmesh/CCNavMesh.cpp]: fixed

File: cocos/navmesh/CCNavMesh.cpp (hunk 2 above)
Comment (songchengjiang) [inline cocos/navmesh/CCNavMesh.cpp]: fixed

File: cocos/navmesh/CCNavMeshAgent.h (hunk 3 above)
Comment (songchengjiang) [inline cocos/navmesh/CCNavMeshAgent.h]: fixed

File: cocos/navmesh/CCNavMeshDebugDraw.cpp (hunk 4 above)
Comment (songchengjiang) [inline cocos/navmesh/CCNavMeshDebugDraw.cpp]: fixed

**Pull request merged**